File: serial.py
# pip install pyserial
import serial
import serial.tools.list_ports

def list_serial_ports():
    ports = serial.tools.list_ports.comports()
    return [port.device for port in ports]

def main():
    ports = list_serial_ports()
    if not ports:
        print("No serial ports found.")
        return

    print("Available serial ports:")
    for p in ports:
        print(f" - {p}")

    port_name = ports[0]
    print(f"Using port: {port_name}")

    try:
        ser = serial.Serial(port=port_name, baudrate=115200, timeout=1)
        print(f"Opened {port_name}")

        # 데이터 송신
        send_data = "$01W10AAEB\n"
        ser.write(send_data.encode('ascii'))
        print(f"Sent: {send_data.strip()}")

        # 데이터 수신 (예: 1초 동안 대기 후 읽기)
        received = ser.read(100)  # 최대 100 바이트 읽기
        if received:
            print("Received:", received.decode('ascii', errors='ignore').strip())
        else:
            print("No data received.")

        ser.close()
        print(f"Closed {port_name}")

    except serial.SerialException as e:
        print(f"Serial error: {e}")

if __name__ == "__main__":
    main()

File: serial.cs
using System;
using System.IO.Ports;
using System.Text;

class SerialComm
{
    private SerialPort _serialPort;

    public SerialComm(string portName, int baudRate = 9600)
    {
        _serialPort = new SerialPort(portName, baudRate, Parity.None, 8, StopBits.One);
        _serialPort.Encoding = Encoding.ASCII;
        _serialPort.DataReceived += SerialPort_DataReceived;
    }

    public void Open()
    {
        if (!_serialPort.IsOpen)
        {
            _serialPort.Open();
            Console.WriteLine($"Serial port {_serialPort.PortName} opened.");
        }
    }

    public void Close()
    {
        if (_serialPort.IsOpen)
        {
            _serialPort.Close();
            Console.WriteLine($"Serial port {_serialPort.PortName} closed.");
        }
    }

    public void Send(string data)
    {
        if (_serialPort.IsOpen)
        {
            _serialPort.Write(data);
            Console.WriteLine($"Sent: {data.Replace("\n", "\\n")}");
        }
        else
        {
            Console.WriteLine("Serial port is not open.");
        }
    }

    private void SerialPort_DataReceived(object sender, SerialDataReceivedEventArgs e)
    {
        try
        {
            string data = _serialPort.ReadExisting();
            Console.WriteLine($"Received: {data.Replace("\n", "\\n")}");
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error in DataReceived handler: " + ex.Message);
        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        // 현재 사용 가능한 포트 검색
        string[] ports = SerialPort.GetPortNames();
        if (ports.Length == 0)
        {
            Console.WriteLine("No serial ports found.");
            return;
        }

        Console.WriteLine("Available serial ports:");
        foreach (var port in ports)
        {
            Console.WriteLine($" - {port}");
        }

        // 첫 번째 포트를 자동 선택
        string portName = ports[0];
        Console.WriteLine($"Using port: {portName}");

        SerialComm serial = new SerialComm(portName, 115200);
        serial.Open();

        serial.Send("$01W10AAEB\n");

        Console.WriteLine("Press any key to exit...");
        Console.ReadKey();

        serial.Close();
    }
}
